### PR TITLE
DOCS: improved documentation

### PR DESCRIPTION
**Pull request by antoinecollet5**
Several changes I propose for the documentation:

Documentation:

-    Use the modern pydata template rather than rtd (used by mainstream packages such as numpy and scipy)
-    Include the readme in the doc for a nice rendering and avoid duplication
-    Make links between the bibliography and the docstrings using the :cite: directive.
-    Improved docstrings for SIES and ES (differentiate class attributes for method parameters, detail inversion types) .
-    Better API section
-    Add makefiles for documentation build (both unix and windows) because tox does not work offline.
-    Add badges in readme for better visibility

### CODE REVIEW

**Comment by tommyod** [comment]
Will look over this more closely soon. :) We just merged a big overhaul of SIES, leading to some conflicts unfortunately. You can probably just merge main and checkout ``--theirs` to solve the conflicts without too much trouble.

**Comment by tommyod** [comment]
Hi. Fantastic work as always! Mostly looks good to me. A few comments:

- You'll have to merge in our recent changes unfortunately.
- We want to avoid makefiles. Can you elaborate on why they are needed if they are? If they are not needed, then we should remove them from the PR.

I'll also try to run it locally before we merge, just to see what the docs look like. But I'll wait until you have merged main :) Good stuff. Thanks for the effort. Greatly appreciate it.

**Comment by antoinecollet5** [comment]
@tommyod thanks for the feedbacks :) I found a bit of time to solve the conflicts and adapt to the new interface. Note that I removed the example in the readme since it is now outdated.

Regarding the  Makefiles, I just removed them.

Wait for your feedback about the html rendering. Note that I did not take care of the pdf export.

File: docs/source/_templates/autosummary/module.rst
Comment (tommyod) [inline docs/source/_templates/autosummary/module.rst:1]: Where is this file from? It is auto-generated from some system? Could you add a reference/source as a comment on the top?

File: docs/source/conf.py
Comment (tommyod) [inline docs/source/conf.py:3]: Remove reference to `pyesmda` please :) Perhaps we should run this setup again, it if it was run in 2017 (many years ago)?

File: docs/source/conf.py
Comment (tommyod) [inline docs/source/conf.py:29]: We have decided on NumPy docstrings, so I think all references to Google docstrings can go away.

File: docs/source/conf.py
Comment (tommyod) [inline docs/source/conf.py:64]: Anything here that we do not need? This list grew quite a bit.

File: docs/source/conf.py
Comment (tommyod) [inline docs/source/conf.py:81]: I note that https://github.com/spatialaudio/nbsphinx has been replaced with https://github.com/executablebooks/MyST-NB. I don't mind, if we have a reason for it.

File: docs/source/conf.py
Comment (tommyod) [inline docs/source/conf.py:85]: No need to suppress warnings if it's not used any more?

File: docs/source/conf.py
Comment (tommyod) [inline docs/source/conf.py:231]: Remove google stuff, since we use NumPy docstrings.

File: docs/source/index.rst
Comment (tommyod) [inline docs/source/index.rst:1]: Very cool! I love the re-use of the README.

File: src/iterative_ensemble_smoother/sies.py
Comment (tommyod) [inline src/iterative_ensemble_smoother/sies.py:31]: I think we should avoid "exposing" attributes in the docs. They are meant for internal use by the class and could in principle change names. Users should interact via the API, not via attributes (unlike e.g. sklearn models, where users are also encouraged to get e.g. `coefs_` from a linear model).

At least that is my thinking. I think exposing attributes leads to more noise, more stuff to maintain, more backwards compatibility issues, etc.

**Review by tommyod — COMMENTED** [review]
Overall very excited about this. New theme looks great. Many improvements here. I do have some comments however :)

New theme looks good:

image


But LaTeX does not render:

image

Adding `"sphinx.ext.mathjax"` to the `extensions` in `conf.py` did not help me locally.

- Could you take a look at LaTeX? Does it work for you?
- I added some comments along the way, please take a look.
- In the future, even smaller PRs is good. For instance, I am not so sure about documenting attributes, but willing to discuss it. However, it's not really related to the issue of updating sphinx etc - so it would've been even better to have a PR for documentation changes (unless they are strictly required to make sphinx run).

File: docs/source/conf.py
Comment (antoinecollet5) [reply docs/source/conf.py:85]: Good catch :)

File: docs/source/_templates/autosummary/module.rst
Comment (antoinecollet5) [reply docs/source/_templates/autosummary/module.rst:4]: The three files in the _template/autosummary/ folder actually define how classes, modules etc. render with autosummary. It is not automatically generated. I can add a readme at the root of the folder if you judge is necessary.

File: docs/source/conf.py
Comment (antoinecollet5) [reply docs/source/conf.py]: Fixed. I cleaned up the conf.py

File: docs/source/conf.py
Comment (antoinecollet5) [reply docs/source/conf.py]: These stuff were simply to rename the "Variable" section to "Attributes" but since you think exposing attributes might not be a good idea, this is not required.

File: docs/source/conf.py
Comment (antoinecollet5) [reply docs/source/conf.py:31]: There were duplicates. I cleaned it.

File: docs/source/conf.py
Comment (antoinecollet5) [reply docs/source/conf.py]: Well... I did not manage to make nbsphinx work. Don't really understand why. I usually rely on this lib. Myst works fine too. I added some config so that the maths render correctly.

File: docs/source/conf.py
Comment (antoinecollet5) [reply docs/source/conf.py]: Removed.

File: docs/source/index.rst
Comment (antoinecollet5) [reply docs/source/index.rst:3]: Cool :)

File: src/iterative_ensemble_smoother/sies.py
Comment (antoinecollet5) [reply src/iterative_ensemble_smoother/sies.py]: Ok I removed it. And you are right, this is anyway  to discuss in another PR anyway.

**Comment by antoinecollet5** [comment]
@tommyod I had a quick look at latex and indeed it does not render. I propose you to see that in another issue/PR ? And to discuss whether the latexpdf export is really needed ?

File: docs/source/_templates/autosummary/module.rst
Comment (tommyod) [reply docs/source/_templates/autosummary/module.rst:4]: If you copied it from some other repo or source, then please add a line saying so. If you wrote it yourself, then no reason to add anything.

**Comment by tommyod** [comment]
> And to discuss whether the latexpdf export is really needed ?

The latexpdf is not needed. But math should render in the HTML. For me it does not render correctly in HTML (see screenshot above, same now as before). Does it render in the HTML docs for you?

I see that [m2r2 expects inline math as](https://github.com/CrossNox/m2r2/issues/20) `` `$e=mc^2$` ``, with the backticks. We use `$e=mc^2$` in our markdown. For blocks of math we use `$$e=mc^2$$` in the notebooks. I am not sure what m2r2 excepts here, or if it supports it at all.

> I propose you to see that in another issue/PR ?

I know it's tedious, but I'd rather get HTML math to work in this PR. I don't want to merge PRs that improve some things, but regress on others, and think that we'll solve it in a future PR. This kind of thinking has a tendency to lead to messy code, especially when it's not obvious what the fix is, who will dedicate time to fix it, etc.

**Comment by antoinecollet5** [comment]
@tommyod normally, math should now render correctly in the html export. Have you tried since yesterday's changes ?

**Comment by tommyod** [comment]
Maybe I did something wrong. Let me try again and report back ... :)

File: docs/source/_templates/autosummary/module.rst
Comment (antoinecollet5) [reply docs/source/_templates/autosummary/module.rst:4]: These three files have been modified from the default templates provided by autosummary. See https://sphinx-rtd-trial.readthedocs.io/en/latest/ext/autosummary.html, section "Customizing templates".
The original files can be found here: https://github.com/sphinx-doc/sphinx/tree/master/sphinx/ext/autosummary/templates/autosummary

I will add the reference in each file.

**Comment by tommyod** [comment]
Fresh try:

```
git remote add antoine https://github.com/antoinecollet5/iterative_ensemble_smoother.git
git fetch --all
git switch feature/docs
pip install -e '.[dev,doc]'
python -m ipykernel install --user
sphinx-build -c docs/source/ -b html docs/source/ docs/build/html -W
firefox docs/build/html/index.html
```

And it works and looks great. :+1:

**Review by tommyod — APPROVED** [review]
LGTM. Thanks again for the work @antoinecollet5 . Very much appreciated as always!

**Pull request merged**
